Sources/DateHelper.swift

```diff
@@ -491,42 +495,91 @@ public extension Date {
-    private static func cachedFormatter(_ format:String = DateFormatType.standard.stringFormat, timeZone: Foundation.TimeZone = Foundation.TimeZone.current, locale: Locale = Locale.current) -> DateFormatter {
-        let hashKey = "\(format.hashValue)\(timeZone.hashValue)\(locale.hashValue)"
-        if Date.cachedDateFormatters[hashKey] == nil {
-            let formatter = DateFormatter()
-            formatter.dateFormat = format
-            formatter.timeZone = timeZone
-            formatter.locale = locale
-            formatter.isLenient = true
-            Date.cachedDateFormatters[hashKey] = formatter
+    internal class concurrentFormatterCache {
+        private static let cachedDateFormattersQueue = DispatchQueue(
+            label: "date-formatter-queue",
+            attributes: .concurrent
+        )
+        
+        private static let cachedNumberFormatterQueue = DispatchQueue(
+            label: "number-formatter-queue",
+            attributes: .concurrent
+        )
+        
+        private static var cachedDateFormatters = [String: DateFormatter]()
+        private static var cachedNumberFormatter = NumberFormatter()
+        
+        private func register(hashKey: String, formatter: DateFormatter) -> Void {
+            concurrentFormatterCache.cachedDateFormattersQueue.async(flags: .barrier) {
+                concurrentFormatterCache.cachedDateFormatters.updateValue(formatter, forKey: hashKey)
+            }
         }
-        return Date.cachedDateFormatters[hashKey]!
-    }
-    
-    /// Generates a cached formatter based on the provided date style, time style and relative date. Formatters are cached in a singleton array using hashkeys.
-    private static func cachedFormatter(_ dateStyle: DateFormatter.Style, timeStyle: DateFormatter.Style, doesRelativeDateFormatting: Bool, timeZone: Foundation.TimeZone = Foundation.NSTimeZone.local, locale: Locale = Locale.current) -> DateFormatter {
-        let hashKey = "\(dateStyle.hashValue)\(timeStyle.hashValue)\(doesRelativeDateFormatting.hashValue)\(timeZone.hashValue)\(locale.hashValue)"
-        if Date.cachedDateFormatters[hashKey] == nil {
-            let formatter = DateFormatter()
-            formatter.dateStyle = dateStyle
-            formatter.timeStyle = timeStyle
-            formatter.doesRelativeDateFormatting = doesRelativeDateFormatting
-            formatter.timeZone = timeZone
-            formatter.locale = locale
-            formatter.isLenient = true
-            Date.cachedDateFormatters[hashKey] = formatter
+        
+        private func retrieve(hashKey: String) -> DateFormatter? {
+            let dateFormatter = concurrentFormatterCache.cachedDateFormattersQueue.sync { () -> DateFormatter? in
+                guard let result = concurrentFormatterCache.cachedDateFormatters[hashKey] else { return nil }
+                
+                return result.copy() as? DateFormatter
```

Comment:
@QuestofIranon Quick question, what's the point of this copy right here?

Reply:
Sorry, I somehow missed this comment. I don't have access to a mac anymore, I suspect my concern was related to avoiding concurrent access to the result by copying it, but I can't tell if it was an actual problem to be concerned about.